# 1. Load the dataset on Colab
# 2. Display the attributes' name and their data type
import pandas as pd
import numpy as np

titanic = pd.read_csv("https://github.com/andvise/DataAnalyticsDatasets/blob/16ca8de1233c8643bfe85fcd1cd87c9ff2221312/titanic.csv?raw=True")





# Part 1: Data Preparation

We import pandas and numpy.

Forewarning: Our code uses slices of the original dataframe and modifies those throughout the program. Colab likes to give a warning about this that I am not modifying the original dataframe and only the copies (the slices). However, I am aware of this and think it's not an issue.

We load the titanic data into pandas, split it and create a min_max_scaler function because we are not allowed to use any libraries except for pandas and numpy.

Survived will be the target label. Because Survived values are binary, we do not need to take precaution with excluding it from the for loop when we run the dataframe features through the min_max_scaler().

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
titanic = pd.read_csv("https://github.com/andvise/DataAnalyticsDatasets/blob/16ca8de1233c8643bfe85fcd1cd87c9ff2221312/titanic.csv?raw=True")
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
882          883         0       2   
883          884         1       1   
884          885         0       3   
885          886         1       1   
886          887         0       3   

                                                  Name     Sex   Age  \
0                               Mr. Owen Harris Braund    male  22.0   
1    Mrs. John Bradley (Florence Briggs Thayer) Cum...  female  38.0   
2                                Miss. Laina Heikkinen  female  26.0   
3          Mrs. Jacques Heath (Lily May Peel) Futrelle  female  35.0   
4                              Mr. William Henry Allen    male  35.0   
..                                                 ...     ...   ...   
882                               Rev. Juozas Montvila    male  27.0   
883                        Miss. Margaret Edith Graham  female  19.0   
884                     Miss. Catherine Helen Johnston  female   7.0   
885                               Mr. Karl Howell Behr    male  26.0   
886                                 Mr. Patrick Dooley    male  32.0   

     Siblings/Spouses Aboard  Parents/Children Aboard     Fare  
0                          1                      0.0   7.2500  
1                          1                      0.0  71.2833  
2                          0                      0.0   7.9250  
3                          1                      NaN  53.1000  
4                          0                      0.0   8.0500  
..                       ...                      ...      ...  
882                        0                      0.0  13.0000  
883                        0                      0.0  30.0000  
884                        1                      2.0  23.4500  
885                        0                      0.0  30.0000  
886                        0                      0.0   7.7500  

[887 rows x 9 columns]

In [ ]:
print(titanic.shape)
print(titanic.dtypes)

(887, 9)
PassengerId                  int64
Survived                     int64
Pclass                       int64
Name                        object
Sex                         object
Age                        float64
Siblings/Spouses Aboard      int64
Parents/Children Aboard    float64
Fare                       float64
dtype: object


In [ ]:
titanic = titanic.drop(labels = "PassengerId", axis = 1)
titanic = titanic.drop(labels = "Name", axis = 1)
titanic

Survived  Pclass     Sex   Age  Siblings/Spouses Aboard  \
0           0       3    male  22.0                        1   
1           1       1  female  38.0                        1   
2           1       3  female  26.0                        0   
3           1       1  female  35.0                        1   
4           0       3    male  35.0                        0   
..        ...     ...     ...   ...                      ...   
882         0       2    male  27.0                        0   
883         1       1  female  19.0                        0   
884         0       3  female   7.0                        1   
885         1       1    male  26.0                        0   
886         0       3    male  32.0                        0   

     Parents/Children Aboard     Fare  
0                        0.0   7.2500  
1                        0.0  71.2833  
2                        0.0   7.9250  
3                        NaN  53.1000  
4                        0.0   8.0500  
..                       ...      ...  
882                      0.0  13.0000  
883                      0.0  30.0000  
884                      2.0  23.4500  
885                      0.0  30.0000  
886                      0.0   7.7500  

[887 rows x 7 columns]

In [ ]:
titanic = titanic.fillna(0)
titanic

Survived  Pclass     Sex   Age  Siblings/Spouses Aboard  \
0           0       3    male  22.0                        1   
1           1       1  female  38.0                        1   
2           1       3  female  26.0                        0   
3           1       1  female  35.0                        1   
4           0       3    male  35.0                        0   
..        ...     ...     ...   ...                      ...   
882         0       2    male  27.0                        0   
883         1       1  female  19.0                        0   
884         0       3  female   7.0                        1   
885         1       1    male  26.0                        0   
886         0       3    male  32.0                        0   

     Parents/Children Aboard     Fare  
0                        0.0   7.2500  
1                        0.0  71.2833  
2                        0.0   7.9250  
3                        0.0  53.1000  
4                        0.0   8.0500  
..                       ...      ...  
882                      0.0  13.0000  
883                      0.0  30.0000  
884                      2.0  23.4500  
885                      0.0  30.0000  
886                      0.0   7.7500  

[887 rows x 7 columns]

In [ ]:
#replacing male with 0 and female with 1. Don't forget which is which!
titanic["Sex"].replace(["male", "female"],
                        [0, 1], inplace=True)

In [ ]:
#data seems to be randomly distributed anyways so no need to take random slices.
#will just slice the last 20% so [710,886] as test and [0,709] as training

train = titanic.iloc[:709, :]
test = titanic.iloc[710:, :]


In [ ]:
#important to normalise or standardise the data at this point since there are 
#differences in the ranges of the data and how they are measured

#define min max scaler function with feature as a parameter
def min_max_scaler(feature):
  scaled_list = []
  for attribute in feature:
    scaled_list.append((attribute - min(feature))/(max(feature) - min(feature)))
  scaled_list
  return(scaled_list)


In [ ]:
for col in train:
  train[col] = min_max_scaler(train[col])

print(train.shape)
train

#We are getting a SettingWithCopyWarning but the code is working so
#we will simply move on for now.

(709, 7)


Survived  Pclass  Sex       Age  Siblings/Spouses Aboard  \
0         0.0     1.0  0.0  0.268139                    0.125   
1         1.0     0.0  1.0  0.470032                    0.125   
2         1.0     1.0  1.0  0.318612                    0.000   
3         1.0     0.0  1.0  0.432177                    0.125   
4         0.0     1.0  0.0  0.432177                    0.000   
..        ...     ...  ...       ...                      ...   
704       1.0     0.0  0.0  0.520505                    0.000   
705       1.0     0.0  1.0  0.268139                    0.000   
706       1.0     1.0  0.0  0.041009                    0.125   
707       1.0     0.0  1.0  0.293375                    0.000   
708       0.0     0.0  0.0  0.507886                    0.000   

     Parents/Children Aboard      Fare  
0                   0.000000  0.014151  
1                   0.000000  0.139136  
2                   0.000000  0.015469  
3                   0.000000  0.103644  
4                   0.000000  0.015713  
..                       ...       ...  
704                 0.000000  0.051310  
705                 0.000000  0.295806  
706                 0.166667  0.029758  
707                 0.000000  0.096626  
708                 0.000000  0.051822  

[709 rows x 7 columns]

In [ ]:
for col in test:
  test[col] = min_max_scaler(test[col])

print(test.shape)
test

(177, 7)


Survived  Pclass  Sex       Age  Siblings/Spouses Aboard  \
710       0.0     1.0  0.0  0.388421                    0.000   
711       0.0     0.5  0.0  0.701006                    0.000   
712       0.0     1.0  0.0  0.252514                    0.000   
713       1.0     0.0  1.0  0.510737                    0.000   
714       1.0     0.5  1.0  0.361239                    0.000   
..        ...     ...  ...       ...                      ...   
882       0.0     0.5  0.0  0.361239                    0.000   
883       1.0     0.0  1.0  0.252514                    0.000   
884       0.0     1.0  1.0  0.089426                    0.125   
885       1.0     0.0  0.0  0.347649                    0.000   
886       0.0     1.0  0.0  0.429193                    0.000   

     Parents/Children Aboard      Fare  
710                      0.0  0.018510  
711                      0.0  0.025374  
712                      0.0  0.014932  
713                      0.0  0.444099  
714                      0.0  0.020495  
..                       ...       ...  
882                      0.0  0.025374  
883                      0.0  0.058556  
884                      0.4  0.045771  
885                      0.0  0.058556  
886                      0.0  0.015127  

[177 rows x 7 columns]

In [ ]:
#We will set the target label as Survived and the rest of the variables as 
#features.

train_data = train.drop("Survived", axis = 1)
train_labels = train["Survived"]

test_data = test.drop("Survived", axis = 1)
test_labels = test["Survived"]

In [ ]:
#look at important features and entropy
print(titanic.groupby("Sex")["Survived"].mean())
print(titanic.groupby("Siblings/Spouses Aboard")["Survived"].mean())

#also perhaps use weighted distances for the more important indicators of 
#survival, e.g. 74% of females survived and 20% of males survived which makes
#it a more important feature than something like Siblings/Spouses Aboard. We
#won't do this immediately but might consider it later on.

Sex
0    0.190227
1    0.742038
Name: Survived, dtype: float64
Siblings/Spouses Aboard
0    0.347682
1    0.535885
2    0.464286
3    0.250000
4    0.166667
5    0.000000
8    0.000000
Name: Survived, dtype: float64


# Part 2 - k-NN implementation


Found example code for calculating Euclidean distance in NumPy from here:

https://www.delftstack.com/howto/numpy/calculate-euclidean-distance/

In [ ]:
test.shape[0]

177

In [ ]:
#we can also find euclidean distance using np.linalg.norm(series1, series2)
#either way is fine

#Euclidean Distance
def Euclidean_distance(a, b):
  distance = np.sqrt(np.sum(np.square(a - b)))
  return(distance)

Euclidean_distance(test.iloc[1,:], train.iloc[1,:])

1.5270611400806646

Found a tutorial online to follow to get the basic structure of a k-NN algorithm:

https://www.youtube.com/watch?v=iJfcRV4PPnY

Needed a mode function for this to work but we are not allowed to use scipy so I
found a mode function to adapt from the internet: 

https://stackoverflow.com/questions/12399107/alternative-to-scipy-mode-function-in-numpy

Edited so that it does not return oldcounts because they are not necessary here.

For our first test we will use a hyperparameter of 3.

In [ ]:
def mode(a, axis=0):
    scores = np.unique(np.ravel(a))
    test_shape = list(a.shape)
    test_shape[axis] = 1
    old_most_freq = np.zeros(test_shape)
    old_counts = np.zeros(test_shape)

    for score in scores:
        template = (a == score)
        counts = np.expand_dims(np.sum(template, axis),axis)
        most_frequent = np.where(counts > old_counts, score, old_most_freq)
        old_counts = np.maximum(counts, old_counts)
        old_most_freq = most_frequent

    return most_frequent

#k-NN algorithm:
def k_NN(train_x, test_x, train_y, test_y, k):
  predictions = []
  for test_pt in test_x.to_numpy():
    distances = []
    for i in range(len(train_x)):
      distances.append(Euclidean_distance(np.array(train_x.iloc[i]), test_pt))

    distance_data = pd.DataFrame(data = distances, columns = ["distance"],
                                 index = train_y.index)
    
    k_neighbours_list = distance_data.sort_values(by = ["distance"],
                                                  axis = 0)[:k]

    labels = train_y.loc[k_neighbours_list.index]

    voting = mode(labels)[0]

    predictions.append(voting)
  return(predictions)

predict_list = k_NN(train_data, test_data, train_labels, test_labels, 3)



In [ ]:
#we have to remember that because we took the slice of test_labels off the end
#of the original dataframe, the indexes start at 710.
def accuracy_score(predict_list, test_labels):
  count = 0
  for score_pos in range(len(test_labels)):
    if test_labels[score_pos + 710] == predict_list[score_pos]:
      count += 1
  accuracy_score = count/(len(test_labels))
  return(accuracy_score)

In [ ]:
accuracy_score(predict_list, test_labels)

0.8587570621468926

In [ ]:
def confusion_matrix(predict_list, test_labels):
  
  true_positives = 0
  false_positives = 0
  true_negatives = 0
  false_negatives = 0

  confusion_matrix = pd.DataFrame({"": ["Predict_Survived", "Predict_Died"],
                                   "Test_Survived": [0,0],
                                   "Test_Died": [0,0]})

  for score_pos in range(len(test_labels)):
    if test_labels[score_pos + 710] == 1:
      if test_labels[score_pos + 710] == predict_list[score_pos]:
        true_positives += 1
      else:
        false_negatives += 1
    else:
      if test_labels[score_pos + 710] == predict_list[score_pos]:
        true_negatives += 1
      else:
        false_positives += 1
    
  confusion_matrix.iloc[0,1] = [true_positives]
  confusion_matrix.iloc[0,2] = [false_positives]
  confusion_matrix.iloc[1,1] = [false_negatives]
  confusion_matrix.iloc[1,2] = [true_negatives]

  return(confusion_matrix)

In [ ]:
confusion_matrix(predict_list, test_labels)

Test_Survived  Test_Died
0  Predict_Survived             53         15
1      Predict_Died             10         99

The confusion matrix reflects that the accuracy score of the k-NN algorithm we have created is high. We can see that the number of true positives (target variable survived both in the test labels and predictions) is 53 and the number of true negatives 99. 53 + 99 / 177 = ~85% which is the accuracy score that we saw earlier. 

# Part 3 - Hyperparameters search 


I decided to use Hyperparameter search again for weighted k_NN later on, so I defined this as a function for ease of use later. The parameters are the list of hyperparameters and the k_NN function we want to use.

In [ ]:
def Hyperparameter_search(hyperparameter_list, k_NN):

  for hyperparameter in HYPERPARAMETERS_LIST:
    predict_list = k_NN(train_data,
                        test_data,
                        train_labels,
                        test_labels,
                        hyperparameter)
    print("Accuracy score and confusion matrix with hyperparameter of "
    + str(hyperparameter) + ": ")
    print("\n")
    print("Accuracy score is: " + str(accuracy_score(predict_list, test_labels)))
    print("\n")
    print("Confusion matrix: ")
    print(confusion_matrix(predict_list, test_labels))
    print("\n" 
          + "---------------------------------------------" 
          + "\n")

HYPERPARAMETERS_LIST = [1, 3, 5, 7, 9, 11]

Hyperparameter_search(HYPERPARAMETERS_LIST, k_NN)

Accuracy score and confusion matrix with hyperparameter of 1: 


Accuracy score is: 0.7909604519774012


Confusion matrix: 
                     Test_Survived  Test_Died
0  Predict_Survived             48         22
1      Predict_Died             15         92

---------------------------------------------

Accuracy score and confusion matrix with hyperparameter of 3: 


Accuracy score is: 0.8587570621468926


Confusion matrix: 
                     Test_Survived  Test_Died
0  Predict_Survived             53         15
1      Predict_Died             10         99

---------------------------------------------

Accuracy score and confusion matrix with hyperparameter of 5: 


Accuracy score is: 0.8700564971751412


Confusion matrix: 
                     Test_Survived  Test_Died
0  Predict_Survived             52         12
1      Predict_Died             11        102

---------------------------------------------

Accuracy score and confusion matrix with hyperparameter of 7: 


Accur

The best k is 5 because it has the highest accuracy score with the most true results in the confusion matrix. We can also see a bit of a trend where as the k tends to 5 it increases (furthest from 5 like 1 and 11 have lower accuracy scores). However, the lowest value besides k == 1 is actually k == 7 which can be seen as an outlier in the trend.

# Part 4 - Weighted k-NN 




I was using a mode approach for normal k-NN, but here it makes more sense to let the higher weight decide which outcome should be predicted. Therefore I modified our code to include positive and negative weight totals using a dataframe that contained the calculated weights and distances.

In [ ]:
print(test_labels.index)

RangeIndex(start=710, stop=887, step=1)


In [ ]:
#weighted k-NN algorithm:
def weighted_k_NN(train_x, test_x, train_y, test_y, k):
  predictions = []
  for test_pt in test_x.to_numpy():
    
    distances = []
    weights = []
    pos_weights_list = []
    neg_weights_list = []
    pos_weight_total = 0
    neg_weight_total = 0

    for i in range(len(train_x)):
      distances.append(Euclidean_distance(np.array(train_x.iloc[i]), test_pt))

    distance_data = pd.DataFrame(data = distances, columns = ["distance"],
                                 index = train_y.index)

    k_neighbours_df = distance_data.sort_values(by = ["distance"])[:k]

    for i in k_neighbours_df["distance"]:
      #code for if i == 0 explained in comments
      if i == 0:
        weights.append(1 / np.square(0.0000001))
      else:
        weights.append(1 / np.square(i))
    
    k_neighbours_df["weights"] = weights

    for i in k_neighbours_df.index:
      if train_y.loc[i] == 1:
        pos_weights_list.append(k_neighbours_df.loc[i, "weights"])
        pos_weight_total = np.sum(pos_weights_list)
      else:
        neg_weights_list.append(k_neighbours_df.loc[i, "weights"])
        neg_weight_total = np.sum(neg_weights_list)
    
    #code for a tie explained in comments
    if pos_weight_total == neg_weight_total:
      if len(pos_weights_list) > len(neg_weights_list):
        predictions.append(0)
      else:
        predictions.append(1)   
    elif pos_weight_total > neg_weight_total:
      predictions.append(1)
    else:
      predictions.append(0)

  return(predictions)

weighted_predict_list = weighted_k_NN(train_data,
                                      test_data,
                                      train_labels,
                                      test_labels, 3)

In [ ]:
accuracy_score(weighted_predict_list, test_labels)

0.8305084745762712

In [ ]:
HYPERPARAMETERS_LIST = [1, 3, 5, 7, 9, 11]

Hyperparameter_search(HYPERPARAMETERS_LIST, weighted_k_NN)

Accuracy score and confusion matrix with hyperparameter of 1: 


Accuracy score is: 0.7909604519774012


Confusion matrix: 
                     Test_Survived  Test_Died
0  Predict_Survived             48         22
1      Predict_Died             15         92

---------------------------------------------

Accuracy score and confusion matrix with hyperparameter of 3: 


Accuracy score is: 0.8305084745762712


Confusion matrix: 
                     Test_Survived  Test_Died
0  Predict_Survived             49         16
1      Predict_Died             14         98

---------------------------------------------

Accuracy score and confusion matrix with hyperparameter of 5: 


Accuracy score is: 0.8361581920903954


Confusion matrix: 
                     Test_Survived  Test_Died
0  Predict_Survived             48         14
1      Predict_Died             15        100

---------------------------------------------

Accuracy score and confusion matrix with hyperparameter of 7: 


Accur

#Comments on results:
It is interesting to note that excluding the case where the hyperparameter is 1 (which of course gives lower accuracy due to not having enough deciding factors), here the accuracy score and confusion matrices stayed nearly exactly the same throughout the testing of multiple hyperparameters. This makes sense as using distance weights to decide the predictions means the size of the k being used is less important. This is because as the hyperparameter grows larger it must include data points further away from the target. However, these points are given less importance because their distance away from the target is factored in.

#In the case of i == 0 
Since i is basically the same point as the target variable it clearly holds weight. To solve the division by 0 problem I have set the value to a miniscule  value of 0.0000001 to give a high weight.

#In the case of a tie:
The list with less weights is given precedence because there may be less weights but they are of a higher value each. For example, there could be a list with 1 weight that has a value of 10 and the other list has 10 weights that each have values of 1. The closest neighbour here is definitely the neighbour with a value of 10 and in my opinion it is more likely to be an accurate predictor of the outcome of the target variable.

#k-NN or Weighted k-NN?

The k-NN algorithm performed better than the weighted k-NN algorithm for this dataset with all values of k that we tested with regards to accuracy score and results of the confusion matrices. Therefore it is perhaps better to use the k-NN algorithm for this specific dataset if we are using a k value of around 5. However, if we are going to use a larger k value then I believe we should use the weighted k-NN algorithm as it has proven to be more constant in accuracy results when it comes to a changing hyperparameter. It is also worth noting that the difference between an 83% accuracy score and a range of [84%,87%] accuracy score is not an extreme difference. Both of the algorithms performed well and produced accurate results.